File: graphs/0200_number_of_islands.py
"""
Problem: Number of Islands
LeetCode #: 200
Difficulty: Medium
Link: https://leetcode.com/problems/number-of-islands/

Approach: Iterate through the grid. When a cell with '1' is encountered, increment the island counter and launch Depth-First Search (DFS) to sink all connected land cells ('1' -> '0').
Time Complexity: O(M * N) where M is rows and N is columns.
Space Complexity: O(M * N) for the worst-case recursion stack.
"""

from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0
        
        rows, cols = len(grid), len(grid[0])
        islands = 0

        def dfs(r: int, c: int) -> None:
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] != '1':
                return
            grid[r][c] = '0'
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    islands += 1
                    dfs(r, c)
                    
        return islands

File: graphs/0200_number_of_islands.java
/*
 * Problem: Number of Islands
 * LeetCode #: 200
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/number-of-islands/
 *
 * Approach: Depth-First Search (DFS) / BFS traversing connected '1' cells and sinking them to '0'.
 * Time Complexity: O(m * n)
 * Space Complexity: O(m * n) recursion stack
 */

public class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        int count = 0;
        int rows = grid.length;
        int cols = grid[0].length;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == '1') {
                    count++;
                    dfs(grid, r, c);
                }
            }
        }
        return count;
    }

    private void dfs(char[][] grid, int r, int c) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length || grid[r][c] != '1') {
            return;
        }

        grid[r][c] = '0'; // Sink island

        dfs(grid, r + 1, c);
        dfs(grid, r - 1, c);
        dfs(grid, r, c + 1);
        dfs(grid, r, c - 1);
    }
}
